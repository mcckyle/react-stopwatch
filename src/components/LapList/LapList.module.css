/* Filename: LapList.module.css */

/* ===============================
   LAP LIST CONTAINER STYLES.
================================== */
.lapList
{
    background: rgba(25, 25, 40, 0.35);
    border-radius: var(--radius);
    backdrop-filter: blur(8px);
    border: 1px solid rgba(255, 255, 255, 0.04);
}

/* =====================
   LAP ROWS STYLES.
======================== */
.lap
{
    display: grid;
    grid-template-columns: 5rem 1fr auto;
    align-items: center;
    padding: 0.65rem 1.1rem;
    border-bottom: 1px solid rgba(255, 255, 255, 0.05);
    font-size: 0.95rem;
    color: var(--text);

    transition: background 0.25s ease, transform 0.2s ease;
}

.lap:hover
{
    background: rgba(255, 255, 255, 0.04);
    transform: translateY(-1px);
}

/* =================================
   LAP LIST MEDIA QUERY STYLES.
==================================== */
@media(max-width: 768px)
{
    .lap
    {
        grid-template-columns: auto auto;
        grid-template-areas:
          "label time"
          "delta delta";
        row-gap: 0.25rem;
    }

    .lapLabel { font-size: 0.75rem; }
    .lapTime { font-size: 0.95rem; }
    .lapDelta { font-size: 0.75rem; }
}
