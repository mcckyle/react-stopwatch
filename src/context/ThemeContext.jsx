//File name: ThemeContext.jsx
//Author: Kyle McColgan
//Date: 7 October 2025
//Description: This file contains the theming context component for the React timer project.

import React, { createContext, useContext, useEffect, useState } from "react";

const ThemeContext = createContext();

export const ThemeProvider = ({ children }) => {
    const getInitialTheme = () =>
      localStorage.getItem("theme") ||
      (window.matchMedia("(prefers-color-scheme: dark)").matches
        ? "dark"
        : "light");

    const [theme, setTheme] = useState(getInitialTheme);

    const toggleTheme = () => {
        setTheme((prev) => (prev === "dark" ? "light" : "dark"));
    };

    useEffect(() => {
        document.documentElement.className = theme;
        localStorage.setItem("theme", theme);
    }, [theme]);

    return (
        <ThemeContext.Provider value={{ theme, toggleTheme }}>
          {children}
        </ThemeContext.Provider>
    );
};

export const useTheme = () => useContext(ThemeContext);
