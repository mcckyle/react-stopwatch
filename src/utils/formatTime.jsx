//File name: formatTime.js
//Author: Kyle McColgan
//Date: 26 October 2025
//Description: This file contains a custom formatting helper function built for the React stopwatch project.

export function formatTime(ms, includeCentiseconds = false)
{
    const centiSeconds = Math.floor((ms % 1000) / 10);
    const seconds = Math.floor((ms / 1000) % 60);
    const minutes = Math.floor((ms / (1000 * 60)) % 60);
    const hours = Math.floor(ms / (1000 * 60 * 60));
    const pad = (num, size = 2) => String(num).padStart(size, "0");

    const formatted = {
        hours: pad(hours),
        minutes: pad(minutes),
        seconds: pad(seconds),
    };

    // return {
    //     hours: pad(hours),
    //     minutes: pad(minutes),
    //     seconds: pad(seconds),
    //     centiSeconds: pad(centiSeconds),
    // };

    if (includeCentiseconds)
    {
        formatted.centiSeconds = pad(centiSeconds);
    }

    return formatted;
}
