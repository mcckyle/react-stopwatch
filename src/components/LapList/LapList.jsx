//File name: LapList.jsx
//Author: Kyle McColgan
//Date: 28 July 2025
//Description: This file contains the Lap component for the React timer site.

import React from "react";
import { motion, AnimatePresence } from "framer-motion";
import styles from "./LapList.module.css";
import { formatTime } from "../../utils/formatTime";
import { useTheme } from "../../hooks/useTheme";

const LapList = ({ laps }) => {

    const theme = useTheme();

    const lapDurations = laps.map((lap, index) => lap - (laps[index + 1] ?? 0));

    const fastest = Math.min(...lapDurations);
    const slowest = Math.max(...lapDurations);

    return (
        <div className = {theme}>
            <div className={styles.lapList}>
              <AnimatePresence mode="popLayout">
                {laps.map((lap, index) => {
                    const lapNumber = laps.length - index;
                    const prevLap = laps[index + 1] ?? 0;
                    const delta = lap - prevLap;
                    const { hours, minutes, seconds, centiSeconds } = formatTime(lap, true);
                    const { minutes: dMin, seconds: dSec, centiSeconds: dCs } = formatTime(delta, true);

                    const deltaLabel =
                      hours !== "00"
                        ? `+${hours}:${minutes}:${seconds}:${centiSeconds}`
                        : `+${dMin}:${dSec}:${dCs}`;

                    let highlight = "";
                    if (delta === fastest) highlight = styles.fastest;
                    else if (delta === slowest) highlight = styles.slowest;

                    return (
                        <motion.div
                          key={index}
                          className={`${styles.lap} ${highlight}`}
                          initial={{ opacity: 0, y: 20, scale: 0.98 }}
                          animate={{ opacity: 1, y: 0, scale: 1 }}
                          exit={{ opacity: 0, y: -10 }}
                          transition ={{
                              duration: 0.4,
                              ease: [0.25, 1, 0.5, 1],
                              delay: index * 0.04,
                          }}
                        >
                            <div className = {styles.lapLabel} aria-label={`Lap ${lapNumber}`}>Lap {lapNumber}</div>
                            <div className = {styles.lapSeparator}> </div>
                            <div className = {styles.lapTime}>
                                {hours !== "00"
                                    ? `${hours}:${minutes}:${seconds}.${centiSeconds}`
                                    : `${minutes}:${seconds}.${centiSeconds}`
                                }
                            </div>
                            {}
                            <div className={styles.lapDelta}>{deltaLabel}</div>
                        </motion.div>
                    );
                })}
              </AnimatePresence>
            </div>
        </div>
    );
};

export default LapList;
