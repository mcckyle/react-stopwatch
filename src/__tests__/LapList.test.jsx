//File name: LapList.test.jsx
//Author: Kyle McColgan
//Date: 15 October 2025
//Description: This file contains the unit test suite for the LapList component.

import React from "react";
import { render, screen } from "@testing-library/react";
import LapList from "../components/LapList/LapList.jsx";

//Mock Framer Motion and dependencies to simplify the animation behavior for testing purposes.
jest.mock("framer-motion", () => ({
    motion: {
        div: ({ children, ...props }) => <div {...props}>{children}</div>,
    },
    AnimatePresence: ({ children }) => <>{children}</>,
}));

//Mock the ThemeContext.jsx
jest.mock("../context/ThemeContext.jsx", () => ({
    useTheme: jest.fn(),
}));

jest.mock("../utils/formatTime", () => ({
    formatTime: jest.fn(),
}));

import { useTheme } from "../context/ThemeContext.jsx";
import { formatTime } from "../utils/formatTime";

describe("LapList Component", () => {

    beforeEach(() => {
        jest.clearAllMocks();
        useTheme.mockReturnValue({ theme: "dark" });
        formatTime.mockImplementation((time, includeCenti) => ({
            hours: "00",
            minutes: "01",
            seconds: "23",
            centiSeconds: "45",
        }));
    });

    //Test #1: Rendering - Renders nothing when there are no laps.
    test("returns null when the laps array is empty", () => {
        const { container } = render(<LapList laps={[]} />);
        expect(container.firstChild).toBeNull();
    });

    //Test #2: Rendering - Renders the correct number of laps.
    test("returns one div per lap entry", () => {
        render(<LapList laps={[1000, 800, 400]} />);
        expect(screen.getAllByText(/Lap/i)).toHaveLength(3);
    });

    //Test #3: Context - Applies the theme class from the context.
    test("applies the theme class to the container", () => {
        const { container } = render(<LapList laps={[1000]} />);
        expect(container.firstChild.className).toContain("dark");
    });

    //Test #4: Data Logic - Displays the lap numbers in reverse order.
    test("renders laps in reverse order (Lap 3 first)", () => {
        render(<LapList laps={[1000, 800, 400]} />);
        const labels = screen.getAllByText(/Lap/i).map(element => element.textContent);
        expect(labels).toEqual(["Lap 3", "Lap 2", "Lap 1"]);
    });

    //Test #5: Function Calls - Calls formatTime correctly for each lap and delta.
    test("calls formatTime with correct arguments for laps", () => {
        render(<LapList laps={[1000, 800, 400]} />);
        expect(formatTime).toHaveBeenCalled();
        //Should be called for lap times AND deltas.
        expect(formatTime).toHaveBeenCalledWith(expect.any(Number), true);
    });

    //Test #6: Display - Displays the formatted lap time correctly.
    test("displays the formatted lap time string", () => {
        render(<LapList laps={[1000]} />);
        expect(screen.getByText("01:23.45")).toBeInTheDocument();
    });


    //Test #7: Display - Displays the formatted delta correctly.
    test("displays formatted delta time with '+' prefix", () => {
        render(<LapList laps={[1000]} />);
        expect(screen.getByText("+01:23.45")).toBeInTheDocument();
    });

    //Test #8: Logic - Highlights the fastest lap.
    test("applies fastest class to lap with smallest delta", () => {
        render(<LapList laps={[1000, 900, 800]} />);

        const lapDivs = screen.getAllByText(/Lap/i).map(lap => lap.parentElement);
        const fastestLap = lapDivs.find(div => div.className.includes("fastest"));

        expect(fastestLap).toBeTruthy();
    });

    //Test #9: Logic - Highlights the slowest lap.
    test("applies slowest class to lap with largest delta", () => {
        render(<LapList laps={[1000, 950, 600]} />);

        const lapDivs = screen.getAllByText(/Lap/i).map(lap => lap.parentElement);
        const slowestLap = lapDivs.find(div => div.className.includes("slowest"));

        expect(slowestLap).toBeTruthy();
    });

    //Test #10: Regression - Snapshot for layout consistency.
    test("matches snapshot", () => {
        const { asFragment } = render(<LapList laps={[1000, 800, 400]} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
