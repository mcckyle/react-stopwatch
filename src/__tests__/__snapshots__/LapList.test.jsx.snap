// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LapList Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="lapList dark"
  >
    <div
      animate="[object Object]"
      class="lap fastest"
      exit="[object Object]"
      initial="[object Object]"
      transition="[object Object]"
    >
      <div
        class="lapLabel"
      >
        Lap 3
      </div>
      <div
        class="lapTime"
      >
        01:23.45
      </div>
      <div
        class="lapDelta"
      >
        +01:23.45
      </div>
    </div>
    <div
      animate="[object Object]"
      class="lap slowest"
      exit="[object Object]"
      initial="[object Object]"
      transition="[object Object]"
    >
      <div
        class="lapLabel"
      >
        Lap 2
      </div>
      <div
        class="lapTime"
      >
        01:23.45
      </div>
      <div
        class="lapDelta"
      >
        +01:23.45
      </div>
    </div>
    <div
      animate="[object Object]"
      class="lap slowest"
      exit="[object Object]"
      initial="[object Object]"
      transition="[object Object]"
    >
      <div
        class="lapLabel"
      >
        Lap 1
      </div>
      <div
        class="lapTime"
      >
        01:23.45
      </div>
      <div
        class="lapDelta"
      >
        +01:23.45
      </div>
    </div>
  </div>
</DocumentFragment>
`;
